### PR TITLE
chore: [LLK-63] LolliPoP login code refactored to share common functionalities

### PR DESCRIPTION
**Pull request by Vangaorth**
⚠ This PR depends on #4443 ⚠

## Short description
This PR refactors the lolliPoP login code to share common functionalities, instead of duplicating them between `IdpLoginScreen` and `CieRequestAuthenticationOverlay`

## List of changes proposed in this pull request
- `ts/features/lollipop/store/reducers/lollipop.ts`: _redux in-memory store_ type edited to use `Option` from _fp-ts_ instead of optionals from _typescript_ (for both `keyTag` and `publicKey`) - state type, reducer and selectors;
- `ts/features/lollipop/store/index.ts`: added store version and migration for _redux-persist_, in order to migrate the persisted `keyTag` from `string?` to `Option<string>`;
- `ts/features/lollipop/hooks/useLollipopLoginSource.tsx`: shared code between `IdpLoginScreen` and `CieRequestAuthenticationOverlay` has been moved here. Removed the `LoginSource` data type since it was redundant;
- `ts/screens/authentication/IdpLoginScreen.tsx`: refactored in order to directly use shared code in `useLollipopLoginSource`;
- `ts/components/cie/CieRequestAuthenticationOverlay.tsx`: refactored in order to directly use shared code in `useLollipopLoginSource`;

## How to test
- Try the lolliPoP persisted store migration: make sure to have a previous application version installed, with a logged lolliPoP user. Update the application version to the one in this PR and make sure that the lolliPoP persisted store is migrated properly (both `keyTag` and `publicKey` have been wrapped in an `Option` and internal values have remained the same)
- Try the login flow, both for Spid and for CIE login, both on dev-server and on production and check that everything works fine. 
- Try the various failure flows and check that they are properly handled (either with a blocking screen or with a retry capability that effectively works)

### CODE REVIEW

File: ts/components/cie/CieRequestAuthenticationOverlay.tsx [hunk 1]
```diff
@@ -97,51 +90,18 @@ const CieWebView = (props: Props) => {
...
-    if (loginSource.kind === "ready") {
-      setWebviewSource(loginSource.value);
-    }
-  }, [loginSource]);
-
-  useEffect(() => {
-    startLoginProcess();
-  }, [startLoginProcess]);
+  const {
+    retryLolliPoPLogin,
```
Comment (shadowsheep1) [inline ts/components/cie/CieRequestAuthenticationOverlay.tsx:101]: I should stay consistent with the friendly term "lollipop" without any assumption about acronyms like PoP or POP as we did from the beginning. What do you think? This gives us also a better **camel case** naming when we use something related to this feature/functionality.

File: ts/components/cie/CieRequestAuthenticationOverlay.tsx [hunk 2]
```diff
@@ -97,51 +90,18 @@ const CieWebView = (props: Props) => {
...
+    shouldBlockUrlNavigationWhileCheckingLolliPoP,
```
Comment (shadowsheep1) [inline ts/components/cie/CieRequestAuthenticationOverlay.tsx:102]: ditto

File: ts/components/cie/CieRequestAuthenticationOverlay.tsx [hunk 3]
```diff
@@ -159,42 +119,8 @@ const CieWebView = (props: Props) => {
...
-        return false;
-      }
-
-      // If code reaches this point, either there is no public key
-      // or lollipop is not enabled or the LolliPOP signature has
-      // been verified (in both cases, let the code flow). Code
-      // flow shall never hit this method is LolliPOP signature
-      // verification has failed, since an error is displayed and
-      // the WebViewSource is left undefined
+    if (shouldBlockUrlNavigationWhileCheckingLolliPoP(url)) {
```
Comment (shadowsheep1) [inline ts/components/cie/CieRequestAuthenticationOverlay.tsx:122]: ditto

File: ts/features/lollipop/hooks/useLollipopLoginSource.tsx [hunk 4]
```diff
@@ -27,27 +35,49 @@ const taskRegenerateKey = (keyTag: string) =>
     TE.chain(() => TE.tryCatch(() => generate(keyTag), toCryptoError))
   );
 
-export const useLollipopLoginSource = (loginUri?: string) => {
-  const [loginSource, setLoginSource] = useState<LoginSourceAsync>({
-    kind: "initial"
-  });
+export const useLollipopLoginSource = (
+  onLolliPoPCheckFailure: () => void,
```
Comment (shadowsheep1) [inline ts/features/lollipop/hooks/useLollipopLoginSource.tsx:39]: ditto

File: ts/features/lollipop/hooks/useLollipopLoginSource.tsx [hunk 5]
```diff
@@ -77,42 +113,119 @@ export const useLollipopLoginSource = (loginUri?: string) => {
...
+  }, [
+    dispatch,
+    loginUri,
+    maybeKeyTag,
+    maybePublicKey,
+    mixpanelEnabled,
+    useLollipopLogin
+  ]);
+
+  const retryLolliPoPLogin = useCallback(() => {
```
Comment (shadowsheep1) [inline ts/features/lollipop/hooks/useLollipopLoginSource.tsx:154]: ditto

File: ts/features/lollipop/hooks/useLollipopLoginSource.tsx [hunk 6]
```diff
@@ -77,42 +113,119 @@ export const useLollipopLoginSource = (loginUri?: string) => {
...
+    // any changes to loginSource otherwise on the next component
+    // refresh (triggered by a different value of loginSource),
+    // the old value of webviewSource is going to be loaded
+    // (i.e., the loaded webViewSource uri will be different from
+    // the loginSource.uri)
+    setWebviewSource(undefined);
+    regenerateLoginSource();
+  }, [regenerateLoginSource]);
+
+  const shouldBlockUrlNavigationWhileCheckingLolliPoP = useCallback(
```
Comment (shadowsheep1) [inline ts/features/lollipop/hooks/useLollipopLoginSource.tsx:166]: ditto

File: ts/screens/authentication/IdpLoginScreen.tsx [hunk 7]
```diff
@@ -251,42 +227,8 @@ const IdpLoginScreen = (props: Props) => {
...
-        return false;
-      }
-
-      // If code reaches this point, either there is no public key
-      // or lollipop is not enabled or the LolliPOP signature has
-      // been verified (in both cases, let the code flow). Code
-      // flow shall never hit this method is LolliPOP signature
-      // verification has failed, since an error is displayed and
-      // the WebViewSource is left undefined
+    if (shouldBlockUrlNavigationWhileCheckingLolliPoP(url)) {
```
Comment (shadowsheep1) [inline ts/screens/authentication/IdpLoginScreen.tsx:230]: ditto

File: ts/features/lollipop/hooks/useLollipopLoginSource.tsx [hunk 8]
```diff
@@ -27,27 +35,49 @@ const taskRegenerateKey = (keyTag: string) =>
...
+          trackLollipopIdpLoginFailure(reason);
+          setLollipopCheckStatus({
+            status: "untrusted",
+            url: O.some(eventUrl)
+          });
+          onLolliPoPCheckFailure();
+        }
+      );
+    },
+    [onLolliPoPCheckFailure]
```
Comment (shadowsheep1) [inline ts/features/lollipop/hooks/useLollipopLoginSource.tsx:77]: ditto

File: ts/features/lollipop/store/index.ts [hunk 9]
```diff
@@ -1,14 +1,45 @@
...
+const migrations: MigrationManifest = {
+  // Version 0
+  // LolliPoP PERSISTED redux type changes from
+  // { keyTag?: string; _persist: ... }
+  // to
+  // { keyTag: O.Option<string>; _persist: ... }
+  "0": (state: PersistedState): PersistedLollipopState => {
+    type PreviousPersistedLolliPoPState = { keyTag?: string } & PersistPartial;
+    const castedPeviousState =
+      state as unknown as PreviousPersistedLolliPoPState;
```
Comment (shadowsheep1) [inline ts/features/lollipop/store/index.ts:26]: ditto

File: ts/screens/authentication/IdpLoginScreen.tsx [hunk 10]
```diff
@@ -125,33 +118,24 @@ const IdpLoginScreen = (props: Props) => {
...
 
   const idpId = props.loggedOutWithIdpAuth?.idp.id;
   const loginUri = idpId ? getIdpLoginUri(idpId) : undefined;
-  const { loginSource, regenerateLoginSource } =
-    useLollipopLoginSource(loginUri);
+  const {
+    retryLolliPoPLogin,
+    shouldBlockUrlNavigationWhileCheckingLolliPoP,
+    webviewSource
+  } = useLollipopLoginSource(handleOnLolliPoPCheckFailure, loginUri);
```
Comment (shadowsheep1) [inline ts/screens/authentication/IdpLoginScreen.tsx:132]: ditto

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]

File: ts/components/cie/CieRequestAuthenticationOverlay.tsx (hunk 1 above)
Comment (Vangaorth) [reply ts/components/cie/CieRequestAuthenticationOverlay.tsx]: Perfect, I do agree with you. Updated

**Review by shadowsheep1 — APPROVED** [review]
LGTM

**Pull request merged**